pallets/cheqd-migration/src/lib.rs

```diff
@@ -0,0 +1,129 @@
+//! A pallet that facilitates the smooth transfer of `DOCK` tokens from the `Dock` Chain to `CHEQD` tokens on the `Cheqd` Chain.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use scale_info::prelude::string::String;
+
+#[cfg(feature = "runtime-benchmarks")]
+pub mod benchmarks;
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+mod weights;
+
+// Re-export pallet items so that they can be accessed from the crate namespace.
+pub use pallet::*;
+
+use weights::{SubstrateWeight, WeightInfo};
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+    use frame_support::{
+        dispatch::WithPostDispatchInfo,
+        pallet_prelude::*,
+        traits::{Currency, ExistenceRequirement::AllowDeath},
+    };
+    use frame_system::pallet_prelude::*;
+    use sp_runtime::traits::Zero;
+
+    type BalanceOf<T> =
+        <<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+    /// Address of the recipient on the `cheqd` side.
+    #[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, Ord, PartialOrd)]
+    #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+    #[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
+    #[derive(scale_info_derive::TypeInfo)]
+    #[scale_info(omit_prefix)]
+    pub struct CheqdAddress(String);
+
+    impl CheqdAddress {
+        pub fn new<T>(value: String) -> Result<Self, Error<T>> {
+            let (prefix, addr) =
+                bech32::decode(&value).map_err(|_| Error::<T>::AddressMustBeValidBech32)?;
+            ensure!(
+                prefix.as_str() == "cheqd",
+                Error::<T>::AddressMustStartWithCheqd
+            );
+            ensure!(addr.len() == 20, Error::<T>::InvalidAddressLength);
+
+            Ok(Self(value))
+        }
+    }
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// The overarching event type.
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+        /// The currency to be burnt during migration.
+        type Currency: Currency<Self::AccountId>;
+        /// Id of the recipient account on the dock side to send burnt funds.
+        type BurnDestination: Get<Self::AccountId>;
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// `cheqd` address must start with `cheqd` prefix.
+        AddressMustStartWithCheqd,
+        /// Provided `cheqd` address is invalid because it has an incorrect length.
+        InvalidAddressLength,
+        /// `cheqd` address part coming after `cheqd` must be a valid bech-32 sequence.
+        AddressMustBeValidBech32,
+        /// Caller account's balance is zero.
+        BalanceIsZero,
+    }
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
```

Comment:
Is there a need of a flag which prevents people from migrating unless we officially start the migration?

Reply:
I thought we could publish the upgrade containing this pallet along with the migration start.